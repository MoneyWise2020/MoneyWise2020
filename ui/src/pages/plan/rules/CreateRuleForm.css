/* https://stackoverflow.com/questions/18146350/how-do-i-simulate-placeholder-functionality-on-input-date-field#:~:text=The%20HTML5%20date%20input%20field%20actually%20does%20not,any%20other%20format%20or%20text%20with%20be%20skipped. */

input[type="date"] {
    position: relative;
}

input[type="date"]:before {
    content: attr(placeholder);
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background: #fff;
    color: rgba(0, 0, 0, 0.65);
    pointer-events: none;
    line-height: 1.5;
    padding: 0 0.5rem;
    /* this differs from StackOverflow answer */
    padding-top: .2rem;
}

input[type="date"]:focus:before,
input[type="date"]:not([value=""]):before {
    display: none;
}