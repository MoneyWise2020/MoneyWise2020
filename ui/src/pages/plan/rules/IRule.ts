// When creating and updating rules
export interface IApiRuleMutate {
    name: string;
    rrule: string;
    value: number;
    labels?: { [label: string]: any };

}

// Extra server-assigned fields which
export interface IApiRule extends IApiRuleMutate {
    id: string;
    userid: string;
    version: string;
}
