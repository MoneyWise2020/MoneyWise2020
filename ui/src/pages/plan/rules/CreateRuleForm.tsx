import React, { useState } from 'react';
import RRule from 'rrule';
import { IApiRuleMutate } from './IRule';

export const CreateForm = ({
    onSubmit,
    onFailedValidation = () => {}
}: {
    onSubmit: (rule: IApiRuleMutate) => void,
    onFailedValidation: (message: string) => void,
}) => {
    const [name, setName] = useState('');
    const [value, setValue] = useState<number | undefined>(undefined);
    const [frequency, setFrequency] = useState<string>("MONTHLY");

    const [bymonthday, setbymonthday] = useState(1);
    const [weekdays, setWeekdays] = useState<string[]>([]);

    const [startDate, setStartDate] = useState<string>('');
    const [endDate, setEndDate] = useState<string>('');
    const [uncertainty, setUncertainty] = useState(false);
    const [highUncertainty, setHighUncertainty] = useState(value);
    const [lowUncertainty, setLowUncertainty] = useState(value);

    function clearForm() {
        setName('');
        setValue(undefined);
        setUncertainty(false);
        setHighUncertainty(0);
        setLowUncertainty(0);
    }
    
    return <form onSubmit={e => {
        e.preventDefault();
        e.stopPropagation();

        if (!value) {
            onFailedValidation(`Please enter a non zero value`);
            return;
        }

        if (name.length < 3) {
            onFailedValidation(`Please enter more than 3 characters, you entered '${name}'`);
            return;
        }

        const dtstart = startDate ? new Date(startDate) : undefined;
        const until = endDate ? new Date(endDate) : undefined;

        let options = {};

        switch (frequency) {
            case "BIWEEKLY":
                if (!dtstart) {
                    onFailedValidation("You must select a start date for 'Biweekly' rules");
                    return;
                }
                options = {
                    freq: RRule.WEEKLY,
                    interval: 2,
                    dtstart,
                    until,
                };
                break;
            case "WEEKLY":
                const weekdayToRRuleWeekdayMapping: { [key: string]: any } = {
                    "SUNDAY": RRule.SU,
                    "MONDAY": RRule.MO,
                    "TUESDAY": RRule.TU,
                    "WEDNESDAY": RRule.WE,
                    "THURSDAY": RRule.TH,
                    "FRIDAY": RRule.FR,
                    "SATURDAY": RRule.SA,
                }
                const rruleWeekdays = weekdays.map(w => weekdayToRRuleWeekdayMapping[w]);
                options = {
                    freq: RRule.WEEKLY,
                    byweekday: rruleWeekdays,
                    dtstart,
                    until,
                };
                break;
            case "MONTHLY":
                options = {
                    freq: RRule.MONTHLY,
                    bymonthday: bymonthday,
                    dtstart,
                    until,
                }
                break;
            case "YEARLY":
                if (!dtstart) {
                    onFailedValidation("You must select a start date for 'Yearly' rules");
                    return;
                }
                options = {
                    freq: RRule.YEARLY,
                    dtstart,
                    until,
                }
                break;
            case "ONCE":
                if (!dtstart) {
                    onFailedValidation("You must select a start date for 'Once' rules");
                    return;
                }
                options = {
                    freq: RRule.YEARLY,
                    dtstart,
                    count: 1,
                };
                break;
        }

        // Success

        const rruleString = new RRule(options).toString()

        onSubmit({
            name: name,
            value: value,
            rrule: rruleString,
            'labels': { 'uncertainty': uncertainty,
                        'highUncertainty': highUncertainty,
                        'lowUncertainty': lowUncertainty
                    },
        });

        clearForm();
    }}>
        <br />
        <br />
        <div className="row">
            <label htmlFor="Name" className="col-sm-2 col-form-label form-control-sm">Name</label>
            <div className="col-sm-8">
                <input className="form-control form-control-sm" id="Name" placeholder="Enter rule name here..." type="text" value={name} onChange={e => setName(e.target.value)} /><br />
            </div>
        </div>

        <div className="row">
        <label htmlFor="Value" className="col-sm-2 col-form-label form-control-sm">Value</label>
        <div className="col-sm-8">
            <input className="form-control form-control-sm" id="Value" type="number" placeholder="Value" step="0.01" value={value} onChange={e => {
                const newValue = Number(e.target.value);
                if (newValue) { // not 0, not undefined
                    setValue(newValue)
                } else {
                    setValue(undefined)
                }
            }} /><br />
        </div>
        </div>

        <div className="form-row">
        <div className="col-md-4 mb-3">
        <label htmlFor="Frequency">Frequency</label>   
            <select className="form-control" id="Frequency" name="frequency" onChange={e => setFrequency(e.target.value)} value={frequency} >
                <option value="WEEKLY">Weekly</option>
                <option value="BIWEEKLY">Biweekly</option>
                <option value="MONTHLY">Monthly</option>
                <option value="YEARLY">Yearly</option>
                <option value="ONCE">Once</option>
            </select>
        </div>

        <div className="col-md-4 mb-3">
        <label htmlFor="Start">Start:</label>
        <input className="form-control" type="date" name="Start" id="Start" value={startDate} onChange={e => setStartDate(e.target.value)} />
        </div>

        {frequency !== "ONCE" && <>
        <div className="col-md-4 mb-3">
            <label htmlFor="End">End:</label>
            <input className="form-control" type="date" name="End" id="End" value={endDate} onChange={e => setEndDate(e.target.value)} />
        </div>
        </>}
        
        {(frequency === "MONTHLY") && <>
        <div className="col-md-4 mb-4">
            <label htmlFor="bymonthday">Day of month</label>
            <input className="form-control" id="bymonthday" type="number" min="1" max="31" placeholder="Day of month" value={bymonthday} onChange={e => setbymonthday(Number(e.target.value))} />
        </div>
        </>}
        
        {(frequency === "WEEKLY") && <>
        <div className="col-md-4 mb-4">
            <label htmlFor="Weekdays">Days of Week</label>
            <select className="custom-select" id="Weekdays" name="Weekdays" multiple value={weekdays} onChange={e => {
                const selectedValues = Array.from(e.target.options)
                    .filter((x) => x.selected)
                    .map((x)=>x.value);

                setWeekdays(selectedValues);
            }}>
                <option value="SUNDAY">Sunday</option>
                <option value="MONDAY">Monday</option>
                <option value="TUESDAY">Tuesday</option>
                <option value="WEDNESDAY">Wednesday</option>
                <option value="THURSDAY">Thursday</option>
                <option value="FRIDAY">Friday</option>
                <option value="SATURDAY">Saturday</option>
            </select></div>
        </>}

        <div className="col-md-4 mb-3">
        <label htmlFor="Uncertainty">Uncertainty:</label>
        <input className="form-control" type="checkbox" name="Uncertainty" id="Uncertainty" checked={uncertainty} onChange={e => setUncertainty(e.target.checked)} />
        </div>

        {(uncertainty) && <>
            <div className="col-md-4 mb-4">
                <label htmlFor="High">High Uncertainty:</label>
                <input className="form-control" type="number" name="High" id="High" value={highUncertainty} onChange={e => setHighUncertainty(Number(e.target.value))} />
                <label htmlFor="Low">Low Uncertainty:</label>
                <input className="form-control" type="number" name="Low" id="Low" value={lowUncertainty} onChange={e => setLowUncertainty(Number(e.target.value))} />
            </div>
        </>}

        </div>
       <button className="btn btn-primary mb-2">Submit</button><br /><br />
    </form>
}
