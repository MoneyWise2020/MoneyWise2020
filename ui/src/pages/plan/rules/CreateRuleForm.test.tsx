import React from 'react';
import { render } from '@testing-library/react'

import {
    setName,
    setValue,
    selectFrequency,
    setDayOfMonth,
    setMonthOfYear,
    setDaysOfWeek,
    setStartDate,
    submit
} from './formUtils.test';

import { CreateForm } from './CreateRuleForm';
import { IApiRuleMutate } from './IRule';

describe('create rule form', () => {
    let submitHandler: jest.MockedFunction<(rule: IApiRuleMutate) => void>;
    let failureHandler: jest.MockedFunction<(message: string) => void>;
    let element: ReturnType<typeof render>;

    beforeEach(() => {
        submitHandler = jest.fn();
        failureHandler = jest.fn();
        element = render(<CreateForm onSubmit={submitHandler} onFailedValidation={failureHandler} />);
    });

    it('should render', () => {
        expect(element).not.toBeUndefined();
        expect(submitHandler).not.toHaveBeenCalled();
    });

    it('should not allow short names', () => {
        setName(element, "1"); // Should be longer name
        setValue(element, -1000);

        selectFrequency(element, "MONTHLY");
        setDayOfMonth(element, 1);

        submit(element);

        expect(submitHandler).not.toHaveBeenCalled();
        expect(failureHandler).toHaveBeenCalledTimes(1);
        expect(failureHandler).toHaveBeenCalledWith(`Please enter more than 3 characters, you entered '1'`);
    });

    it('should not allow 0 value', () => {
        setName(element, "Rent");
        setValue(element, 0);

        selectFrequency(element, "MONTHLY");
        setDayOfMonth(element, 1);

        submit(element);

        expect(submitHandler).not.toHaveBeenCalled();
        expect(failureHandler).toHaveBeenCalledTimes(1);
        expect(failureHandler).toHaveBeenCalledWith(`Please enter a non zero value`);
    });

    describe("weekly", () => {
        it('should submit for multiple days', () => {
            setName(element, "Gas");
            setValue(element, -25);
    
            selectFrequency(element, "WEEKLY");
            setDaysOfWeek(element, ['MONDAY', 'TUESDAY']);
    
            submit(element);
    
            expect(failureHandler).not.toHaveBeenCalled();
            expect(submitHandler).toHaveBeenCalledTimes(1);
    
            const rule = submitHandler.mock.calls[0][0];
            expect(rule).toEqual(expect.objectContaining({
                name: 'Gas',
                value: -25
            }));
            expect(rule.rrule).toMatchSnapshot();
        });
    });

    describe("biweekly", () => {
        it('should submit', () => {
            setName(element, "Paycheck");
            setValue(element, 1800);
    
            selectFrequency(element, "BIWEEKLY");
            setStartDate(element, 2020, 10, 10);

            submit(element);
    
            expect(failureHandler).not.toHaveBeenCalled();
            expect(submitHandler).toHaveBeenCalledTimes(1);
    
            const rule = submitHandler.mock.calls[0][0];
            expect(rule).toEqual(expect.objectContaining({
                name: 'Paycheck',
                value: 1800,
            }));
            expect(rule.rrule).toMatchSnapshot();
        });

        it('should not submit without start date', () => {
            setName(element, "Paycheck");
            setValue(element, 1800);
    
            selectFrequency(element, "BIWEEKLY");

            submit(element);
    
            expect(submitHandler).not.toHaveBeenCalled();
            expect(failureHandler).toHaveBeenCalledTimes(1);
            expect(failureHandler).toHaveBeenCalledWith("You must select a start date for 'Biweekly' rules");
        });
    });

    describe("monthly", () => {
        it('should submit', () => {
            setName(element, "Rent");
            setValue(element, -1000);
    
            selectFrequency(element, "MONTHLY");
            setDayOfMonth(element, 15);
    
            submit(element);
    
            expect(failureHandler).not.toHaveBeenCalled();
            expect(submitHandler).toHaveBeenCalledTimes(1);
    
            const rule = submitHandler.mock.calls[0][0];
            expect(rule).toEqual(expect.objectContaining({
                name: 'Rent',
                value: -1000
            }));
            expect(rule.rrule).toMatchSnapshot();
        });
    });

    describe("yearly", () => {
        it('should submit', () => {
            setName(element, "Birthday Present!");
            setValue(element, -42);
    
            selectFrequency(element, "YEARLY");
            setMonthOfYear(element, 10);
            setDayOfMonth(element, 1);
    
            submit(element);
    
            expect(failureHandler).not.toHaveBeenCalled();
            expect(submitHandler).toHaveBeenCalledTimes(1);
    
            const rule = submitHandler.mock.calls[0][0];
            expect(rule).toEqual(expect.objectContaining({
                name: 'Birthday Present!',
                value: -42
            }));
            expect(rule.rrule).toMatchSnapshot();
        });
    });

    describe("once", () => {
        // TODO(jamesfulford): Why is this not passing?
        xit('should submit', () => {
            setName(element, "Haaaaahvaaaahd Tuition");
            setValue(element, -2900);
    
            selectFrequency(element, "ONCE");
            setStartDate(element, 2020, 11, 8);
    
            submit(element);
    
            expect(failureHandler).not.toHaveBeenCalled();
            expect(submitHandler).toHaveBeenCalledTimes(1);
    
            const rule = submitHandler.mock.calls[0][0];
            expect(rule).toEqual(expect.objectContaining({
                name: 'Haaaaahvaaaahd Tuition',
                value: -2900
            }));
            expect(rule.rrule).toMatchSnapshot();
        });
    });
});
