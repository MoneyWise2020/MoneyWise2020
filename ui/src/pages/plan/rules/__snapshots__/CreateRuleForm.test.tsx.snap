// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create rule form biweekly should submit 1`] = `
"DTSTART:20201010T000000Z
RRULE:FREQ=WEEKLY;INTERVAL=2"
`;

exports[`create rule form monthly should submit 1`] = `"RRULE:FREQ=MONTHLY;BYMONTHDAY=15"`;

exports[`create rule form weekly should submit for multiple days 1`] = `"RRULE:FREQ=WEEKLY;BYDAY=MO,TU"`;

exports[`create rule form yearly should submit 1`] = `
"DTSTART:20201010T000000Z
RRULE:FREQ=YEARLY"
`;
